src/test/java/io/confluent/examples/streams/window/CustomWindowTest.java

```diff
@@ -0,0 +1,254 @@
+        // this integration test's timeout (30 secs) to ensure we observe the expected processing results.
+        streamsConfiguration.put(StreamsConfig.COMMIT_INTERVAL_MS_CONFIG, 10 * 1000);
+        streamsConfiguration.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+        // Use a temporary directory for storing state, which will be automatically removed after the test.
+        streamsConfiguration.put(StreamsConfig.STATE_DIR_CONFIG, TestUtils.tempDirectory().getAbsolutePath());
+
+        final KStream<String, Integer> input = builder.stream(inputTopic);
+        final KTable<Windowed<Integer>, Integer> sumOfOddNumbers = input
+                .selectKey((k, v) -> 1)
+                .groupByKey()
+                .windowedBy(new DailyTimeWindows(zoneId, Duration.ofMinutes(30)))
+                // A simple sum of value
+                .reduce((v1, v2) -> v1 + v2, Materialized.with(Serdes.Integer(), Serdes.Integer()))
+                // We only care about final result
+                .suppress(Suppressed.untilWindowCloses(Suppressed.BufferConfig.unbounded()));
+        sumOfOddNumbers.toStream().print(Printed.toSysOut());
+        sumOfOddNumbers.toStream().to(outputTopic, Produced.with(WindowedSerdes.timeWindowedSerdeFrom(Integer.class), Serdes.Integer()));
+
+        TopologyTestDriver testDriver = new TopologyTestDriver(builder.build(), streamsConfiguration);
+
+        injectFakeData(inputValues, testDriver);
+        verifyResults(expectedValues, testDriver);
+
+        testDriver.close();
+    }
+
+    private class MyEvent {
+        Integer value;
+        ZonedDateTime timestamp;
+
+        public MyEvent(Integer value, ZonedDateTime timestamp) {
+            this.value = value;
+            this.timestamp = timestamp;
+        }
+    }
+
+    private class ExpectedResult {
+        Integer key;
+        Integer value;
+        ZonedDateTime start;
+        ZonedDateTime end;
+
+        public ExpectedResult(Integer key, Integer value, ZonedDateTime start, ZonedDateTime end) {
+            this.key = key;
+            this.value = value;
+            this.start = start;
+            this.end = end;
+        }
+
+        public Windowed<Integer> toWindowed() {
+            return new Windowed<>(key, new TimeWindow(start.toInstant().toEpochMilli(), end.toInstant().toEpochMilli()));
+        }
+    }
+
+    private void verifyResults(List<ExpectedResult> expectedValues, TopologyTestDriver testDriver) {
+        for (ExpectedResult expectedValue : expectedValues) {
+            ProducerRecord<Windowed<Integer>, Integer> results = testDriver.readOutput(outputTopic, new TimeWindowedDeserializer(new IntegerDeserializer(), Duration.ofDays(1).toMillis()), new IntegerDeserializer());
+            OutputVerifier.compareKeyValue(results, expectedValue.toWindowed(), expectedValue.value);
+
+        }
+    }
+
+    private void injectFakeData(List<MyEvent> inputValues, TopologyTestDriver testDriver) {
```

Comment:
There are now some utilities in IntegrationTestUtils that you could use for this and for `verifyResults`.

Reply:
I tried but, it doesn't support providing event with given timestamp yet :)